zerver/migrations/0458_remove_invalid_characters_from_user_group_name.py

```diff
@@ -0,0 +1,77 @@
+# Generated by Django 4.2.2 on 2023-07-03 06:39
+
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def check_group_name_starts_with_invalid_prefix(group_name: str) -> str:
+    INVALID_NAME_PREFIXES = ["@", "role:", "user:", "stream:", "channel:"]
+
+    for invalid_prefix in INVALID_NAME_PREFIXES:
+        if group_name.startswith(invalid_prefix):
+            return invalid_prefix
+    return ""
+
+
+def remove_invalid_characters_from_user_group_name(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    # Remove "@" character from name of role-based system groups.
+    UserGroup = apps.get_model("zerver", "UserGroup")
+    Realm = apps.get_model("zerver", "Realm")
+
+    groups_to_update = []
+    INVALID_NAME_PREFIXES = ["@", "role:", "user:", "stream:", "channel:"]
+    for group in UserGroup.objects.filter(is_system_group=True):
+        if group.name[0] != "@":
+            # We skip the group if its name has already been updated.
+            continue
+
+        old_group_name = group.name
+        group.name = old_group_name[1:]
+        groups_to_update.append(group)
+
+    # Fix the name of non-system groups as well.
+    for realm in Realm.objects.all():
+        existing_group_names = UserGroup.objects.filter(
+            realm=realm, is_system_group=False
+        ).values_list("name", flat=True)
+        existing_group_names_set = set(existing_group_names)
+        for prefix in INVALID_NAME_PREFIXES:
+            groups = UserGroup.objects.filter(
+                realm=realm, is_system_group=False, name__startswith=prefix
+            )
+            for group in groups:
+                group_name = group.name
+                while True:
+                    matching_invalid_prefix = check_group_name_starts_with_invalid_prefix(
+                        group_name
+                    )
+                    if len(matching_invalid_prefix) == 0:
+                        break
+                    group_name = group_name[len(matching_invalid_prefix) :]
```

Comment:
This `while` loop is to make sure that we handle cases like `@role:admin` or `@@role:admin`.

Reply:
We can do `group_name = group_name.lstrip(matching_invalid_prefix)`.

Reply:
I tried using it initially but we cannot, `lstrip` strips all the characters and not a string. For example - if `group_name = stream:admin` and I call `group_name.lstrip("stream:")`, the result will be `dmin` because `stream:` contains `a` character and it is also removed.

Reply:
Ahh, I see. Thanks for explaining.